compiler/semexprs.nim

```diff
@@ -2153,6 +2153,25 @@ proc semMagic(c: PContext, n: PNode, s: PSym, flags: TExprFlags): PNode =
     result[0] = newSymNode(s, n[0].info)
     result[1] = semAddrArg(c, n[1], s.name.s == "unsafeAddr")
     result.typ = makePtrType(c, result[1].typ)
+  of mBitandI:
```

Comment:
Too hacky. I understand why you need to do it this way but I wonder if there is a better way...

Reply:
Tried to fix this line instead:

```nim

  # semfold.nim
  of mBitandI, mAnd: result = newIntNodeT(bitand(a.getInt, b.getInt), n, g)
```

Or to add logic to ``semmagic.magicsAfterOverloadResolution``? Sem'checking both operands for ``and`` and ``or`` does make sense.

Reply:
> and or 

good point, `or` should also short-circuit, I will update PR

But for the remaining points:

How could that work? if you semcheck the 2nd operand for `and` , `const a = false and nonexistant` would not compile; nor would `when declared(Foo) and T is Foo: discard`

`Error: undeclared identifier` would be issued before code reaches semmagic.magicsAfterOverloadResolution or semfold.evalOp.semcheck, which is why I'm doing it instead in `semMagic`

If you `nim c -r main` with:
```nim
static:echo "ok1"
const a = false and nonexistant
```
and you add
```nim
# semfold.nim
  of mBitandI, mAnd: 
    echo ("semfold", g.config$a.info,)
    ...
```

and
```nim
# semmagic
proc magicsAfterOverloadResolution(c: PContext, n: PNode, flags: TExprFlags): PNode =
  echo ("magicsAfterOverloadResolution", c.config$n.info,)
  ...
```

you'll see:
```
ok1
main.nim: Error: undeclared identifier: 'nonexistant'
```

Reply:
> How could that work? if you semcheck the 2nd operand for and , const a = false and nonexistant would not compile; nor would when declared(Foo) and T is Foo: discard
> Error: undeclared identifier would be issued before code reaches

Exactly. As it should be. I'm sorry, I seem to have misunderstood this point previously. Short cut evaluation refers to *evaluation*, not to the semantic checking of ``and`` and ``or``.

Reply:
you seemed to agree here https://github.com/nim-lang/Nim/issues/13502#issuecomment-591341784 it should be legal
> This is a bug, not a feature request. In fact, it's a duplicate of some other bug report.

feature or bug is a matter of perspective, here's a motivating example from https://github.com/nim-lang/Nim/pull/13498/files#diff-7eb4d1654e7ba69a150250a34148c9c9

```nim
template ctAnd(a, b): bool =
  # pending https://github.com/nim-lang/Nim/issues/13502
  when a:
    when b: true
    else: false
  else: false

template initImpl(result: typed, size: int) =
  when ctAnd(declared(SharedTable), type(result) is SharedTable):
    init(result, size)
  else: ...
```

or, without ctAnd you'd need to duplicate the else branch (worst option) or write boilerplate:
```nim
when declared(SharedTable):
  template fallback() = ...
  when type(result) is SharedTable:
    init(result, size)
  else:
    fallback()
  else:
    fallback()
```

with this PR you'd simply write:
```nim
template initImpl(result: typed, size: int) =
  when declared(SharedTable) and type(result) is SharedTable:
    init(result, size)
  else: ...
```

which is arguably simpler. Are you implying https://github.com/nim-lang/Nim/issues/13502 should be closed as wontfix? I don't see downsides and only upsides to making `when declared(Foo) and T is Foo` work.

Reply:
> which is arguably simpler.

Well it's simpler to use but a weird special case.

Reply:
note that I ensured it only affects static/when/const blocks, not general VM code, so that:
```nim
template fun() = # ditto with generics, procs etc
  let a = b and c # uses full semcheck despite running inside VM
  const a2 = b and c # uses short-circuit
static: fun()
```
seems reasonable feature, if feature is accepted I can make sure doc is clear

Reply:
Sorry, but I'm really happier without this special case, regardless of "only const/when/static blocks" are affected or not.

Reply:
would you accept `lazyAnd`, `lazyOr` (please suggest better name) in stdlib
* system?
* sugar?
* elsewhere?

```
template lazyAnd(lhs: bool, rhs: untyped): bool =
  when lhs: rhs
  else: false
```
and analog for lazyOr

Reply:
As I said, do not circumvent the RFC process. ``lazyAnd`` is a good name, but consider that your very own implementation is a single when statement, why not write it as this when the need arises:

```nim

when (when declared(SharedTable): (typeof(result) is SharedTable) else: false):
  code here
```